File: src/js/Vector.ts
export default class Vector{
  x: number
  y: number

  constructor(x?: number, y?: number){
    this.x = x || 0;
    this.y = y || 0;
  }

  abs(): Vector{
    return new Vector(Math.abs(this.x), Math.abs(this.y));
  }

  add(add: Vector|number): Vector{
    if(typeof add === 'number')
      return new Vector(this.x + add, this.y + add);
    else
      return new Vector(this.x + add.x, this.y + add.y);
  }

  clone(): Vector{
    return new Vector(this.x, this.y);
  }

  cross(vector: Vector): number{
    return this.x * vector.y - this.y * vector.x;
  }

  dist(vector: Vector): number{
    return Math.sqrt(this.distSq(vector));
  }

  distSq(vector: Vector): number{
    const dx = this.x - vector.x;
    const dy = this.y - vector.y;

    return dx * dx + dy * dy;
  }

  div(div: Vector|number): Vector{
    if(typeof div === 'number')
      return new Vector(this.x / div, this.y / div);
    else
      return new Vector(this.x / div.x, this.y / div.y);
  }

  dot(vector: Vector): number{
    return this.x * vector.x + this.y * vector.y;
  }

  equals(vector: Vector): boolean{
    return vector.x === this.x && vector.y === this.y;
  }

  length(): number{
    return Math.sqrt(this.dot(this));
  }

  lengthSq(): number{
    return this.dot(this);
  }

  lerp(vector: Vector, s: number): Vector{
    return new Vector(this.x + (vector.x - this.x) * s, this.y + (vector.y - this.y) * s);
  }

  mul(mul: Vector|number): Vector{
    if(typeof mul === 'number')
      return new Vector(this.x * mul, this.y * mul);
    else
      return new Vector(this.x * mul.x, this.y * mul.y);
  }

  normalize(): Vector{
    return this.div(this.length());
  }

  set(set: Vector|number): Vector{
    if(typeof set === 'number'){
      this.x = set;
      this.y = set;
    }
    else{
      this.x = set.x;
      this.y = set.y;
    }

    return this;
  }

  setLength(length: number): Vector{
    return this.normalize().mul(length);
  }

  sub(sub: Vector|number): Vector{
    if(typeof sub === 'number')
      return new Vector(this.x - sub, this.y - sub);
    else
      return new Vector(this.x - sub.x, this.y - sub.y);
  }

  truncate(max: number): Vector{
    if(this.length() > max)
      return this.normalize().mul(max);
    else
      return this;
  }

  reverse(): Vector{
    return this.mul(-1);
  }

}

File: src/cpp/nbody/vector.cc
#include "vector.h"

#include <cmath>

namespace nbody {

// NOTE: This adaption of the Vector class is excessively functional for 
// the purposes of a high-performance n-body simulation. This is only 
// a provisional starting point.

Vector::Vector(float x, float y) {
  x_ = x;
  y_ = y;
}

Vector Vector::Abs() const {
  return Vector(fabs(x_), fabs(y_));
}

Vector Vector::Add(float addend) const {
  // possible overflow
  return Vector(x_ + addend, y_ + addend);
}

Vector Vector::Add(const Vector& addend) const {
  // possible overflow
  return Vector(x_ + addend.x_, y_ + addend.y_);
}

Vector Vector::Clone() const {
  return Vector(x_, y_);
}

float Vector::Cross(const Vector& vector) const {
  // possible overflow
  return (x_ * vector.y_ - y_ * vector.x_);
}

float Vector::Distance(const Vector& vector) const {
  return sqrt(DistanceSquared(vector));
}

float Vector::DistanceSquared(const Vector& vector) const {
  // possible underflow and overflow
  float dx = x_ - vector.x_;
  float dy = y_ - vector.y_;
  return dx * dx + dy * dy;
}

Vector Vector::Divide(float divisor) const {
  // possible div by zero
  return Vector(x_ / divisor, y_ / divisor);
}

Vector Vector::Divide(const Vector& divisor) const {
  // possible div by zero
  return Vector(x_ / divisor.x_, y_ / divisor.y_);
}

float Vector::Dot(const Vector& vector) const {
  // possible overflow
  return (x_ * vector.x_ + y_ * vector.y_);
}

bool Vector::Equals(const Vector& vector) const {
  return (vector.x_ == x_ && vector.y_ == y_);
}

float Vector::Length() const {
  return sqrt(LengthSquared());
}

float Vector::LengthSquared() const {
  return Dot(this);
}

Vector Vector::Lerp(const Vector& vector, float amount) const {
  // possible underflow and overflow
  // TODO: check that this is correct interpolation (by the definition).
  return Vector(x_ + (vector.x_ - x_) * amount, y_ + (vector.y_ - y_) * amount);
}

Vector Vector::Multiply(Vector& multiplier) const {
  // possible overflow
  return Vector(x_ * multiplier.x_, y_ * multiplier.y_);
}

Vector Vector::Multiply(float multiplier) const {
  // possible overflow
  return Vector(x_ * multiplier, y_ * multiplier);
}

Vector Vector::Normalize() const {
  return Divide(Length());
}

void Vector::Set(float target) {
  x_ = target;
  y_ = target;
}

void Vector::Set(Vector target) {
  x_ = target.x_;
  y_ = target.y_;
}

// DANGER: this function (which was likely written incorrectly from the 
// original Vector.ts) did not originally mutate to the new length. 
// Instead, it returned Normalize().Multiply().
void Vector::SetLength(float length) {
  Vector temp = Normalize().Multiply(length);
  x_ = temp.x_;
  y_ = temp.y_;
}

Vector Vector::Subtract(float sub) const {
  // possible underflow
  return Vector(x_ - sub, y_ - sub);
}

Vector Vector::Subtract(Vector sub) const {
  // possible underflow
  return Vector(x_ - sub.x_, y_ - sub.y_);
}

void Vector::Truncate(float max_length) {
  if (Length() > max_length)
    SetLength(max_length);
}

float Vector::Dot(const Vector* vector) const {
  // possible overflow
  return (x_ * vector->x_ + y_ * vector->y_);
}

} // namespace nbody
